dimod/include/dimod/quadratic_model.h

```diff
@@ -1094,6 +849,123 @@ class BinaryQuadraticModel : public QuadraticModelBase<Bias, Index> {
     Vartype vartype_;
 };
 
+template<class T>
+struct VarInfo {
+    Vartype vartype;
+    T lb;
+    T ub;
+
+    VarInfo(Vartype vartype, T lb, T ub) : vartype(vartype), lb(lb), ub(ub) {}
+};
+
+template <class Bias, class Index = int>
+class QuadraticModel : public QuadraticModelBase<Bias, Index> {
+ public:
+    /// The type of the base class.
+    using base_type = QuadraticModelBase<Bias, Index>;
+
+    /// The first template parameter (Bias).
+    using bias_type = Bias;
+
+    /// The second template parameter (Index).
+    using index_type = Index;
+
+    /// Unsigned integral that can represent non-negative values.
+    using size_type = typename base_type::size_type;
+
+    QuadraticModel() : base_type(), varinfo_() {}
+
+    template <class B, class T>
+    explicit QuadraticModel(const BinaryQuadraticModel<B, T>& bqm)
+            : base_type(bqm) {
+        if (bqm.vartype() == Vartype::SPIN) {
+            this->varinfo_.insert(this->varinfo_.begin(), bqm.num_variables(),
+                                  VarInfo<bias_type>(Vartype::SPIN, -1, 1));
+        } else if (bqm.vartype() == Vartype::BINARY) {
+            this->varinfo_.insert(this->varinfo_.begin(), bqm.num_variables(),
+                                  VarInfo<bias_type>(Vartype::BINARY, 0, 1));
+        } else {
+            throw std::logic_error("unexpected vartype");
+        }
+    }
+
+    void add_quadratic(index_type u, index_type v, bias_type bias) {
+        if (u == v) {
+            assert(0 <= u && static_cast<size_t>(u) < this->num_variables());
```

Comment:
Same as before. I generally write a macro for these asserts. If they are going to appear too many times in a file.

#define ASSERT_CHECK_VARIABLE(var_type, v) \\
    assert(v >= 0 && v < this->num_variables(var_type))

Reply:
I like the explicitness of putting the actual assert in.